src/ring_theory/dedekind_domain/adic_valuation.lean

```diff
@@ -222,4 +239,222 @@ begin
+by { rw valuation_of_algebra_map, exact v.int_valuation_lt_one_iff_dvd r }
+
+/-- The `v`-adic valuation of `0 : K` equals 0. -/
+lemma valuation.map_zero' : v.valuation_def (0 : K) = 0 :=
+begin
+  rw [← (algebra_map R K).map_zero, valuation_of_algebra_map v],
+  exact v.int_valuation.map_zero',
+end
+
+/-- The `v`-adic valuation of `1 : K` equals 1. -/
+lemma valuation.map_one' : v.valuation_def (1 : K) = 1 :=
+begin
+  rw [← (algebra_map R K).map_one, valuation_of_algebra_map v],
+  exact v.int_valuation.map_one',
+end
+
+/-- The `v`-adic valuation of a product is the product of the valuations. -/
+lemma valuation.map_mul' (x y : K) :
+  v.valuation_def (x * y) = v.valuation_def x * v.valuation_def y :=
+begin
+  rw [valuation_def, valuation_def, valuation_def, div_mul_div_comm₀, ← int_valuation.map_mul',
+    ← int_valuation.map_mul', ← submonoid.coe_mul],
+  apply int_valuation_div_eq K v,
+  rw [(classical.some_spec (valuation_def._proof_2 (x * y))), is_fraction_ring.mk'_eq_div,
+    (algebra_map R K).map_mul, submonoid.coe_mul, (algebra_map R K).map_mul, ← div_mul_div_comm₀,
+    ← is_fraction_ring.mk'_eq_div, ← is_fraction_ring.mk'_eq_div,
+    (classical.some_spec (valuation_def._proof_2 x)),
+    (classical.some_spec (valuation_def._proof_2 y))],
+end
+
+/-- The `v`-adic valuation of a sum is bounded above by the maximum of the valuations. -/
+lemma valuation.map_add_le_max' (x y : K) :
+  v.valuation_def (x + y) ≤ max (v.valuation_def x) (v.valuation_def y) :=
+begin
+  obtain ⟨rx, sx, hx⟩ := is_localization.mk'_surjective (non_zero_divisors R) x,
+  obtain ⟨rxy, sxy, hxy⟩ := is_localization.mk'_surjective (non_zero_divisors R) (x + y),
+  obtain ⟨ry, sy, hy⟩ := is_localization.mk'_surjective (non_zero_divisors R) y,
+  rw [← hxy, ← hx, ← hy, valuation_of_mk', valuation_of_mk', valuation_of_mk'],
+  have h_frac_xy : is_localization.mk' K rxy sxy =
+    is_localization.mk' K (rx*(sy : R) + ry*(sx : R)) (sx*sy),
+  { rw [is_localization.mk'_add, hx, hy, hxy], },
+  have h_frac_x : is_localization.mk' K rx sx = is_localization.mk' K (rx*(sy : R)) (sx*sy),
+  { rw [is_localization.mk'_eq_iff_eq, submonoid.coe_mul, mul_assoc, mul_comm (sy : R) _], },
+  have h_frac_y : is_localization.mk' K ry sy = is_localization.mk' K (ry*(sx : R)) (sx*sy),
+  { rw [is_localization.mk'_eq_iff_eq, submonoid.coe_mul, mul_assoc], },
+  have h_denom : 0 < v.int_valuation_def ↑(sx * sy),
+  { rw [int_valuation_def, if_neg _],
+    { exact with_zero.zero_lt_coe _ },
+    { exact non_zero_divisors.ne_zero
+        (submonoid.mul_mem (non_zero_divisors R) sx.property sy.property), }},
+  rw [int_valuation_div_eq K v h_frac_x, int_valuation_div_eq K v h_frac_y,
+    int_valuation_div_eq K v h_frac_xy, le_max_iff, div_le_div_right₀ (ne_of_gt h_denom),
+    div_le_div_right₀ (ne_of_gt h_denom), ← le_max_iff],
+  exact v.int_valuation.map_add_le_max' _ _,
+end
+
+/-- The `v`-adic valuation on `K`. -/
+def valuation  : valuation K (with_zero (multiplicative ℤ)) :=
```

Comment:
I'm fairly sure a similar construction generalizes to any situation where `v` is a valuation on a domain `A` whose support is the zero ideal. I wonder whether it would be better to make such a general construction and specialize here using `int_valuation`.

Reply:
OK, I'll try to generalize it. Thanks!

Reply:
I opened the following pull request with such a construction (well, slightly more general): #13610

Reply:
Thanks, #13610 looks good to me! I'll wait until that is merged and the uniform space diamond is solved to continue working on this PR.